Backport the cluster scripts to 2.1

ctubbsii opened the pull request:
This backports the improvements to the accumulo-cluster and accumulo-service scripts to 2.1 to improve script maintenance and functionality across the branches.

Notable differences with the main branch:

* This version includes deprecated commands that are replaced by newer versions: start-non-tservers, start-servers, stop-servers, start-tservers, and stop-tservers
* This vesion includes the ability to control starting a compaction-coordinator, which was removed in the main branch for 4.0
* This version does not allow specifying tserver groups (unsupported by its cluster.yaml version) and uses a hard-coded value of "default"
* As before these changes, this version checks for old deprecated config files present in conf/
* As before these changes, this version parses the cluster.yaml that was supported for the 2.1 and 3.1 branches
* This version is modified to translate the variables emitted from its version of the ClusterConfigParser to the newer names used in the main branch
* This version does not require compactor groups to be configured (since external compactions are optional)
* This version uses `-a <host>` instead of `-o general.process.bind.address=<host>` for the bind address
* This version uses `-q` and `-g` for the group parameter

This includes relevant changes from #4966 to standardize the service names, #5066 to reduce ssh connections to hosts, #5114 to support starting/stopping groups of servers by type, #5126 to support resource group overrides, #5116 to refactor and simplify the script argument parsing using getopt, and the bugfixes and improvements from #5167

ctubbsii commented:
The diffs here are probably not going to be very helpful. The more relevant diff is between these changed files and the current contents of 3.1 and main. Some of these differences go away when merging into 3.1, and most of the rest goes away when merging into main. There are a few small improvements in here that I do expect to include in the main branch, though. Those are specifically:

```patch
diff --git a/assemble/bin/accumulo-cluster b/assemble/bin/accumulo-cluster
index a87ad4938a..31dff57d81 100755
--- a/assemble/bin/accumulo-cluster
+++ b/assemble/bin/accumulo-cluster
@@ -38,7 +38,7 @@ Commands:
   start-here                 Deprecated. Alias for "start --local --all"
   stop-here                  Deprecated. Alias for "stop --local --all"
   restart [--local] [--all | [--manager] [--gc] [--monitor] [--tservers[=group]] [--sservers[=group]] [--compactors[=group]]
-                             Restarts the Accumulo cluster
+                             Restarts Accumulo cluster services
   start [--local] [--all | [--manager] [--gc] [--monitor] [--tservers[=group]] [--sservers[=group]] [--compactors[=group]]
                              Starts Accumulo cluster services
   stop [--local] [--all | [--manager] [--gc] [--monitor] [--tservers[=group]] [--sservers[=group]] [--compactors[=group]]
@@ -53,7 +53,7 @@ Commands:
   accumulo-cluster start --local                           # start all local services
   accumulo-cluster start --local --manager                 # start local manager services
   accumulo-cluster start --tservers                        # start all tservers
-  accumulo-cluster start --tservers=group1                 # start all group1 tservers
+  accumulo-cluster start --sservers=group1                 # start all group1 sservers
   accumulo-cluster start --local --manager --tservers      # Start the local manager and local tservers
 
 EOF
@@ -140,7 +140,7 @@ function parse_args {
   if [[ $ARG_ALL == 1 && ($ARG_MANAGER == 1 ||
     $ARG_GC == 1 || $ARG_MONITOR == 1 || $ARG_TSERVER == 1 ||
     $ARG_SSERVER == 1 || $ARG_COMPACTOR == 1) ]]; then
-    echo "--all cannot be used with other options"
+    echo -- "--all cannot be used with other options"
     print_usage
     exit 1
   fi
@@ -152,20 +152,17 @@ function parse_args {
     ARG_ALL=1
   fi
 
-  if isDebug; then
-    echo "DEBUG=$DEBUG"
-    echo "ARG_ALL=$ARG_ALL"
-    echo "ARG_LOCAL=$ARG_LOCAL"
-    echo "ARG_MANAGER=$ARG_MANAGER"
-    echo "ARG_GC=$ARG_GC"
-    echo "ARG_MONITOR=$ARG_MONITOR"
-    echo "ARG_TSERVER=$ARG_TSERVER"
-    echo "ARG_TSERVER_GROUP=$ARG_TSERVER_GROUP"
-    echo "ARG_SSERVER=$ARG_SSERVER"
-    echo "ARG_SSERVER_GROUP=$ARG_SSERVER_GROUP"
-    echo "ARG_COMPACTOR=$ARG_COMPACTOR"
-    echo "ARG_COMPACTOR_GROUP=$ARG_COMPACTOR_GROUP"
-  fi
+  debug "ARG_ALL=$ARG_ALL"
+  debug "ARG_LOCAL=$ARG_LOCAL"
+  debug "ARG_MANAGER=$ARG_MANAGER"
+  debug "ARG_GC=$ARG_GC"
+  debug "ARG_MONITOR=$ARG_MONITOR"
+  debug "ARG_TSERVER=$ARG_TSERVER"
+  debug "ARG_TSERVER_GROUP=$ARG_TSERVER_GROUP"
+  debug "ARG_SSERVER=$ARG_SSERVER"
+  debug "ARG_SSERVER_GROUP=$ARG_SSERVER_GROUP"
+  debug "ARG_COMPACTOR=$ARG_COMPACTOR"
+  debug "ARG_COMPACTOR_GROUP=$ARG_COMPACTOR_GROUP"
 }
 
 function invalid_args {
diff --git a/assemble/bin/accumulo-service b/assemble/bin/accumulo-service
index ae5c424de8..e56e647227 100755
--- a/assemble/bin/accumulo-service
+++ b/assemble/bin/accumulo-service
@@ -111,7 +111,7 @@ function start_service() {
     rotate_log "$outfile"
     rotate_log "$errfile"
 
-    nohup "${bin}/accumulo" "$service_type" "$@" "${PROPERTY_OVERRIDES[@]}" >"$outfile" 2>"$errfile" </dev/null &
+    nohup "${bin}/accumulo" "$service_type" "$@" >"$outfile" 2>"$errfile" </dev/null &
     echo "$!" >"${pid_file}"
 
   done
```

ddanielr reviewed (commented):
Added definitions for the external compaction services caused an error. 

Section added to cluster.yaml 
```
compaction:
  coordinator:
    - localhost
  compactor:
    - q1:
        - localhost
    - q2:
        - localhost
        
tservers_per_host: 2
sservers_per_host:
 - default: 1
compactors_per_host:
 - q1: 1
 - q2: 1
```

Error:
```
$./bin/accumulo-cluster start --dry-run
DEBUG: args:  --dry-run --
DEBUG: ARG_ALL=1
DEBUG: ARG_LOCAL=0
DEBUG: ARG_MANAGER=0
DEBUG: ARG_GC=0
DEBUG: ARG_MONITOR=0
DEBUG: ARG_COORDINATOR=0
DEBUG: ARG_TSERVER=0
DEBUG: ARG_TSERVER_GROUP=
DEBUG: ARG_SSERVER=0
DEBUG: ARG_SSERVER_GROUP=
DEBUG: ARG_COMPACTOR=0
DEBUG: ARG_COMPACTOR_GROUP=
DEBUG: LOCAL_HOST_ADDRESSES=::1 10-113-15-93 10.113.15.93 127.0.0.1 172.17.0.1 192.168.49.1 ip-10-113-15-93.evoforge.org ip-172-17-0-1.ec2.internal ip-192-168-49-1.ec2.internal localhost localhost4 localhost4.localdomain4 localhost6 localhost6.localdomain6 localhost.localdomain
DEBUG: Parsed config:
MANAGER_HOSTS="localhost"
MONITOR_HOSTS="localhost"
GC_HOSTS="localhost"
TSERVER_HOSTS="localhost"
COORDINATOR_HOSTS="localhost"
COMPACTION_QUEUES="q1 q2"
COMPACTOR_HOSTS_q1="localhost"
NUM_COMPACTORS_q1="1"
COMPACTOR_HOSTS_q2="localhost"
NUM_COMPACTORS_q2="1"
SSERVER_GROUPS="default"
SSERVER_HOSTS_default="localhost"
NUM_SSERVERS_default="1"
NUM_TSERVERS="${NUM_TSERVERS:=2}"
starting tablet servers for group default
DEBUG: ssh -qnf -o ConnectTimeout=2 localhost bash\ -c\ \'ACCUMULO_CLUSTER_ARG=2\ \"/workspace/fluo-uno/install/accumulo-2.1.4-SNAPSHOT/bin/accumulo-service\"\ tserver\ start\ -a\ localhost\ \' 
done
DEBUG: ssh -qnf -o ConnectTimeout=2 localhost bash\ -c\ \'ACCUMULO_CLUSTER_ARG=1\ \"/workspace/fluo-uno/install/accumulo-2.1.4-SNAPSHOT/bin/accumulo-service\"\ manager\ start\ -a\ localhost\ \' 
DEBUG: ssh -qnf -o ConnectTimeout=2 localhost bash\ -c\ \'ACCUMULO_CLUSTER_ARG=1\ \"/workspace/fluo-uno/install/accumulo-2.1.4-SNAPSHOT/bin/accumulo-service\"\ gc\ start\ -a\ localhost\ \' 
DEBUG: ssh -qnf -o ConnectTimeout=2 localhost bash\ -c\ \'ACCUMULO_CLUSTER_ARG=1\ \"/workspace/fluo-uno/install/accumulo-2.1.4-SNAPSHOT/bin/accumulo-service\"\ monitor\ start\ -a\ localhost\ \' 
./bin/accumulo-cluster: line 365: COMPACTION-COORDINATORS_PER_HOST_default: invalid variable name
```

Removed all the cluster changes and only adding a definition for the compaction-coordinator triggered the error. 
```
compaction:
  coordinator:
    - localhost
```

ddanielr reviewed (changes requested):
These changes don't seem to be working for the external compaction functionality

ctubbsii commented:
Ah, right, I thought we might run into this. The bash variable is based on the service name, and the coordinator has a dash in the service name, which you can't have in a bash variable. I'll work around that.

dlmarion commented on assemble/conf/accumulo.properties:
```diff
@@ -31,3 +31,8 @@ instance.secret=DEFAULT
 tserver.memory.maps.native.enabled=true
+
+## (optional) include additional property files for a resource group
+## based on the ACCUMULO_RESOURCE_GROUP env var set in accumulo-service
```
Yeah, in 2.1 and 3.1, it's only applicable for compactors and scan servers.

ctubbsii replied:
This is backported word-for-word from the main branch. Nothing in the comment says anything about which server types the accumulo-service script will set it for. In fact, accumulo-cluster and accumulo-service aren't really well documented at all, regarding what kinds of environment variables they make use of or set when launching processes. I don't think the shortcoming is the docs for this comment, but rather the lack of docs on accumulo-service in general. That's not something that I intend to solve with this PR, though.

Also, note that tservers do get the "default" resource group, just like any other service type that does not have the ability to specify resource groups in cluster.yaml (manager, gc, coordinator, monitor). This is also implied by the contents of accumulo-env.sh

If this comment were to change, I'd drop the mention of accumulo-service entirely, and simply say something like: "the following lines demonstrate the ability to organize properties into different files, named according to an environment variable; see commons-configuration documentation for details on property interpolation and include directives". After all, the commented out examples are not specific to accumulo-cluster or accumulo-service, or any other script. It's just a general feature that users can take advantage of if they have scripts that set environment variables.

ddanielr replied:
Fair enough.  I tested adding a tserver resource group in the cluster.yaml for `default` and got a parsing error: `Processing error: Unknown configuration section : tserver.default`. 

That's a bit more direct than a code comment.

ctubbsii replied:
Yeah, I didn't change any of the cluster yaml parsing (other than use a more specific exception type and remove unneeded return statements, which was a backport of the same changes in 3.1), so none of the yaml features have changed at all, just the script internals and the related accumulo-env.sh handling of the emitted script environment variables.

dlmarion commented on assemble/bin/accumulo-cluster:
```diff
@@ -27,598 +35,517 @@ Options:
+  ARG_COMPACTOR=0
+  ARG_COMPACTOR_GROUP=""
+
+  PARSE_OUTPUT=$(getopt -o "" --long "dry-run,all,local,manager,gc,monitor,compaction-coordinator,no-tservers,tservers,sservers::,compactors::" -n 'accumulo-cluster' -- "$@")
```
I don't see `no-tservers` in the usage.

ctubbsii replied:
And by "For simplicity", I mean it was nearly impossible to parse the option out otherwise. It had to be added to getopt somehow, and the logic for only being applicable for `start`, but only when `start` was called via `start-tservers` was a bit too complicated, and completely unnecessary, since the deprecated functionality could still be supported by just adding it to all the commands.

dlmarion commented on assemble/bin/accumulo-cluster:
```diff
@@ -27,598 +35,517 @@ Options:
+  debug "ARG_MONITOR=$ARG_MONITOR"
+  debug "ARG_COORDINATOR=$ARG_COORDINATOR"
+  debug "ARG_TSERVER=$ARG_TSERVER"
+  debug "ARG_TSERVER_GROUP=$ARG_TSERVER_GROUP"
```
`ARG_TSERVER_GROUP` does not apply in 2.1 and 3.1

ctubbsii replied:
That's why it's never set when getopt options are parsed. It remains empty. This is just script internals to keep it aligned with what's in the main branch. It doesn't mean tserver groups are supported.

If you're looking at these changes by comparing with what's in 2.1, I suggest instead comparing it to what's in main. It's a much smaller diff, and makes a lot more sense.

ddanielr commented:
@ctubbsii found two issues when testing changes in 4.0. 
They are pretty minor so I created a PR against your branch so they would get pulled in.
https://github.com/ctubbsii/accumulo/pull/22

ctubbsii commented:
> @ctubbsii found two issues when testing changes in 4.0. They are pretty minor so I created a PR against your branch so they would get pulled in. [ctubbsii#22](https://github.com/ctubbsii/accumulo/pull/22)

I'm not sure what you mean by "testing changes in 4.0". These changes are the backport of what's in the main/4.0 branch. Are you saying those changes are things that are also problems in the main branch, and should be fixed here, so they get merged up into main/4.0?

ddanielr commented:
> > @ctubbsii found two issues when testing changes in 4.0. They are pretty minor so I created a PR against your branch so they would get pulled in. [ctubbsii#22](https://github.com/ctubbsii/accumulo/pull/22)
> 
> I'm not sure what you mean by "testing changes in 4.0". These changes are the backport of what's in the main/4.0 branch. Are you saying those changes are things that are also problems in the main branch, and should be fixed here, so they get merged up into main/4.0?

Yes that's exactly what I am saying. :+1:

ddanielr commented on assemble/conf/accumulo-env.sh:
```diff
@@ -89,23 +89,33 @@ JAVA_OPTS=(
+    esac
+    ;;
+  *)
+    echo "ACCUMULO_RESOURCE_GROUP named $ACCUMULO_RESOURCE_GROUP is not configured"
```
This message shows up when running `./accumulo-cluster` but it's not apparent where the configuration needs to be updated. 

```suggestion
    echo "ACCUMULO_RESOURCE_GROUP named $ACCUMULO_RESOURCE_GROUP is not configured in accumulo-env.sh."
```

ctubbsii commented:
@dlmarion and @ddanielr anything else before I merge? I think I've addressed all the things (but not sure if it's to everybody's satisfaction)

ddanielr commented:
Yeah, I think you're good to merge

ddanielr commented:
Yeah, I think you're good to merge

dlmarion commented:
@ctubbsii  - no issues with you merging. @ddanielr and I discussed these changes yesterday, and although I didn't get a chance to test them out I know that he did.

ctubbsii commented:
In these latest changes, I added colorized output to make it prettier, and fixed a few things with argument parsing. I noticed that `--all` had some problems, so I just removed it. I think it's better without. But it can be added back in pretty easily if we want to... we just need to decide on more concrete behavior for it. There's also somethings we can do to lock down repeated options, like calling different servers types with different groups... because right now, the latest one takes precedence. (Previous to these changes, the latest would take precedence, *except* if you specified all servers of that type by omitting the group name, in which case, the earlier one that had the group name present would take precedence, which was very confusing and weird).

Also, there's a lot of logic that's intertwined with accumulo-service, and accumulo-service is clobbering the console with things that the accumulo-cluster script was trying to make nice.... like accumulo-service now prints a bunch of stuff that gets in the way of the progress dots in accumulo-cluster for starting tservers. Also, if we're going to do the progress dots, it should probably be done generally, not just for tservers. (like maybe scan servers and compactors, too?)

dlmarion commented:
`shellcheck` is failing in `main`, but not the other branches.

ctubbsii commented:
> `shellcheck` is failing in `main`, but not the other branches.

Fixed in 3015c1b25e3f80336ac1177edf7e4b21936ba05b
